ce/api/geo.py

```diff
@@ -52,88 +53,144 @@ def inner(obj, _seen_ids = None):
 
-        with cache_lock:
-            self.cache[key] = result
-            self.misses += 1
-            while getsize(self.cache) > self.maxsize * 1024 * 1024: # convert to MB
-                self.cache.popitem(0) # Purge least recently used cache entry
+                return result
 
-        return result
+            #cache testing / debugging functions
+            def cache_clear(self):
+                with cache_lock:
+                    self.cache.clear()
+                    self.hits = 0
+                    self.misses = 0
 
-    def cache_clear(self):
-        with cache_lock:
-            self.cache.clear()
-            self.hits = 0
-            self.misses = 0
+            def get_hits(self): return self.hits
 
-@memoize_mask
-def wkt_to_masked_array(nc, fname, wkt, variable):
-    poly = loads(wkt)
-    return polygon_to_masked_array(nc, fname, poly, variable)
+            def get_misses(self): return self.misses
 
+            def get_size(self): return getsize(self.cache)
 
-def polygon_to_masked_array(nc, fname, poly, variable):
+            def get_length(self): return len(self.cache)
+
+        cf = CachedFunction(self.maxsize, self.keyfunc, func)
+        return cf
 
+def make_mask_grid_key(nc, fname, poly, variable):
+    '''
+    Generates a key for characterizing a numpy mask  meant to
+    be applied to netCDF files: the polygon the mask is generated from
+    and min / max / number of steps for both latitutde and longitude.
+    Assumes a regular grid.
+    '''
+    latsteps = nc.variables['lat'].shape[0]
+    latmin = nc.variables['lat'][0]
+    latmax = nc.variables['lat'][latsteps - 1]
+    lonsteps = nc.variables['lon'].shape[0]
+    lonmin = nc.variables['lon'][0]
+    lonmax = nc.variables['lon'][lonsteps - 1]
+    wkt = dumps(poly) #dict-style geoJSON can't be hashed
+    return (wkt, latmin, latmax, latsteps, lonmin, lonmax, lonsteps)
+
+@memoize(make_mask_grid_key, 10)
+def polygon_to_mask(nc, fname, poly, variable):
+    """Generates a numpy mask from a polygon"""
     nclons = nc.variables['lon'][:]
     if np.any(nclons > 180):
         poly = translate(poly, xoff=180)
 
     dst_name = 'NETCDF:"{}":{}'.format(fname, variable)
     with rasterio.open(dst_name, 'r', driver='NetCDF') as raster:
-
-        if raster.affine == rasterio.Affine.identity():
+        if raster.transform == rasterio.Affine.identity():
             raise Exception("Unable to determine projection parameters for GDAL "
                             "dataset {}".format(dst_name))
+        mask, out_transform, window = rio_getmask(raster, [mapping(poly)], all_touched=True)
+
+    return mask, out_transform, window
 
-        the_array, _ = rio_mask(raster, [mapping(poly)], crop=False, all_touched=True)
+def make_masked_file_key(nc, fname, wkt, varname):
+    """generates a key suitable for characterizing a masked netCDF file:
+       filename and polygon"""
+    return (fname, wkt)
+
+@memoize(make_masked_file_key, 100)
```

Comment:
So I would propose that we make the maxsize argument here parameterizable as an application option. This is something that we'll probably want to change based on the deployment. 100MB might be too much for working on our laptops, but we could probably afford more on  larger deploy with dedicated resources. Having some flexibility available from the top level would be nice. Let's add that as a feature request... doesn't need to be part of this PR.